File: lenses-alerts-java/src/main/java/io/lenses/alerting/plugin/javaapi/AlertingPlugin.java
/*
 * Copyright 2017-2019 Lenses.io Ltd
 */
package io.lenses.alerting.plugin.javaapi;

import io.lenses.alerting.plugin.javaapi.util.Try;

import java.util.List;
import java.util.Map;
import java.util.Optional;

public interface AlertingPlugin {

    /**
     * Provides the name for the alert integration service. For Slack integration this will read Slack.
     * @return The alert integration name
     */
    String name();

    /**
     * Provides information on what and how this alert integration service works.
     * @return
     */
    String description();

    /**
     * Provides a list of all the configuration keys
     * @return
     */
    List<ConfigEntry> configKeys();

    /**
     * Returns AlertingService for valid configuration (depends on the alerting service class).
     * @param config service configuration
     * @return alerting service or failure
     */
    Try<AlertingService> init(Map<String, String> config);

    /**
     * Returns AlertingService for valid configuration (depends on the alerting service class).
     * @param config service configuration
     * @param nameOverride override for the name of the service
     * @param descriptionOverride override for the description of the service
     *
     * @return alerting service or failure
     */
    Try<AlertingService> init(Map<String, String> config, Optional<String> nameOverride, Optional<String> descriptionOverride);
}

File: lenses-alerts-scala/src/main/scala/io/lenses/alerting/plugin/scalaapi/AlertingPlugin.scala
/*
 * Copyright 2017-2019 Lenses.io Ltd
 */
package io.lenses.alerting.plugin.scalaapi

import java.util.Optional

import io.lenses.alerting.plugin.javaapi.{AlertingPlugin => JAlertingPlugin}
import io.lenses.alerting.plugin.scalaapi.util.TryUtils._

import scala.collection.JavaConverters._
import scala.util.Try

/**
  * simple alerting plugin wrapper
  */
class AlertingPlugin(underlying: JAlertingPlugin) {

  def init(config: Map[String, String], nameOverride: Option[String], descriptionOverride: Option[String]): Try[AlertingService] = for {
    init <- Try {
      underlying.init(config.asJava, Optional.ofNullable(nameOverride.orNull), Optional.ofNullable(descriptionOverride.orNull))
    }
    service <- init.asScala
  } yield new AlertingService(service)

  def name: String = underlying.name()

  def description: String = underlying.description()

  def configKeys: Seq[ConfigEntry] = underlying.configKeys().asScala.map { c => ConfigEntry(c.getKey, c.getDescription) }.toList
}

object AlertingPlugin {
  def apply(underlying: JAlertingPlugin): AlertingPlugin = new AlertingPlugin(underlying)
}

case class ConfigEntry(key: String, desc: String)
